### DISCUSSION TITLE
Imported many binaries with wrong dsPIC language. How to fix quickly?

### DISCUSSION THREAD

**Discussion by gtackett** [NationalSecurityAgency/ghidra#2729] [Q&A]
I accidentally imported a boat load (thousands) of binaries with the wrong dsPIC33*x* selected.

These binaries are arranged correctly into 3 different folders, one for each flavor of dsPIC33.

I just need a quick way to set all the binaries in one folder to one selected architecture.

I could script this, but I'm not very clear on how to handle prompting for the new architecture. Once I have that, I already know how to script applying it to all the binaries in a folder..

Maybe someone else here can help?

**Comment by astrelsky** [comment]
`GhidraScript` has an `askLanguage` method. This will provide the prompt for the new architecture.

**Reply by gtackett** [reply]
I did know about ```askLanguage```, but thanks for mentioning it. What I'm having trouble getting to work is ```setLanguage```.

I'm doing this from Jython, and unfortunately, I notice now that ```ghidra.program.database.ProgramDB``` has two methods ```setLanguage``` which differ in their first argument: One expects a ```LanguageTranslator``` object but the other (the one I want) expects a ```Language``` object.

I believe there's no way from Jython to specify which of the two I actually want?

If there isn't, maybe there's a way I can get a ```LanguageTranslator``` from the program's current ```Language``` to its new one. But this appears to take some more work with the API that I haven't figured out yet.

**Reply by astrelsky** [reply]
Jython should automatically choose the correct method depending on the type.

**Reply by astrelsky** [reply]
You could use [LanguageTranslatorFactory](https://ghidra.re/ghidra_docs/api/ghidra/program/util/LanguageTranslatorFactory.html) to get a translator from the languages if needed. There would need to be a `*.trans` file though.

**Reply by gtackett** [reply]
> Jython should automatically choose the correct method depending on the type.
```

>>> currentProgram.setLanguage(askLanguage('Choose','OK'),currentProgram.getCompiler(),False,monitor)
Traceback (most recent call last):
  File "python", line 1, in <module>
TypeError: setLanguage(): 1st arg can't be coerced to ghidra.program.model.lang.Language, ghidra.program.util.LanguageTranslator
```

**Reply by astrelsky** [reply]
Should be `currentProgram.setLanguage(askLanguage('Choose','OK').getLanguage(),currentProgram.getCompiler(),False,monitor)` as `askLanguage` returns a `LanguageCompilerSpecPair`

**Reply by gtackett** [reply]
That gives me the same TypeError message.

This can wait until next week, if you're getting bored of it right now. 😄

**Reply by astrelsky** [reply]
> That gives me the same TypeError message.
> 
> This can wait until next week, if you're getting bored of it right now. 😄

Faulty error message. It's the second argument that is the problem.

```python
lang = askLanguage('Choose','OK')
id = lang.compilerSpec.compilerSpecID
lang = lang.language
currentProgram.setLanguage(lang, id, False, monitor)
```

If this results in an `IncompatibleLanguageException` then you may need to implement the `*.trans` file or a `LanguageTranslatorAdapter`.

Also, if you want to do this in a headless script you can hardcode [LanguageCompilerSpecPair(String, String)](https://ghidra.re/ghidra_docs/api/ghidra/program/model/lang/LanguageCompilerSpecPair.html#%3Cinit%3E(java.lang.String,java.lang.String))

**Comment by gtackett** [comment]
It's going to require less hard work to just re-import everything, even though it may take an hour or so to run.

**Reply by astrelsky** [reply]
That's a shame. There really should be a default for this when translating between variants.
